File: src/main/kotlin/dev/murad/shipping/network/SetEnginePacket.java
package dev.murad.shipping.network;

import io.netty.buffer.ByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;

import static dev.murad.shipping.network.VehiclePacketHandler.LOCATION;

public record SetEnginePacket(int locoId, boolean state) implements CustomPacketPayload {

    public static final Type<SetEnginePacket> TYPE = new Type<>(LOCATION);

    // Each pair of elements defines the stream codec of the element to encode/decode and the getter for the element to encode
    // 'name' will be encoded and decoded as a string
    // 'age' will be encoded and decoded as an integer
    // The final parameter takes in the previous parameters in the order they are provided to construct the payload object
    public static final StreamCodec<ByteBuf, SetEnginePacket> STREAM_CODEC = StreamCodec.composite(
            ByteBufCodecs.VAR_INT,
            SetEnginePacket::locoId,
            ByteBufCodecs.BOOL,
            SetEnginePacket::state,
            SetEnginePacket::new
    );

    @Override
    public Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }
}

File: src/main/kotlin/dev/murad/shipping/network/SetEnginePacket.kt
package dev.murad.shipping.network

import dev.murad.shipping.HumVeeMod
import io.netty.buffer.ByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.ResourceLocation
import java.util.function.BiFunction

data class SetEnginePacket(val locoId: Int, val state: Boolean) : CustomPacketPayload {

    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload?> {
        return TYPE
    }

    companion object {

        private val LOCATION = ResourceLocation.fromNamespaceAndPath(HumVeeMod.MOD_ID, "locomotive_channel_engine_packet")

        val TYPE: CustomPacketPayload.Type<SetEnginePacket> = CustomPacketPayload.Type<SetEnginePacket>(LOCATION)
        
        val STREAM_CODEC: StreamCodec<ByteBuf?, SetEnginePacket> =
            StreamCodec.composite<ByteBuf, SetEnginePacket, Int, Boolean>(
                ByteBufCodecs.VAR_INT,
                SetEnginePacket::locoId,
                ByteBufCodecs.BOOL,
                SetEnginePacket::state
            ) { locoId, state -> SetEnginePacket(locoId, state) }
    }
}
